core/templates/dev/head/services/assets-backend-api.service.spec.ts

```diff
@@ -108,20 +147,158 @@ describe('Assets Backend API Service', function() {
+      successHandler, failHandler);
+    $httpBackend.flush();
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalledWith('myfile.mp3');
+    $httpBackend.verifyNoOutstandingExpectation();
+  });
+
+  it('should handler rejection on fetching file when Blob throws a' +
+    ' TypeError and BlobBuilder does not work correctly', function() {
+    var successHandler = jasmine.createSpy('success');
+    var failHandler = jasmine.createSpy('fail');
+
+    spyOn(window, 'Blob').and.callFake(function() {
+      throw new TypeError();
+    });
+
+    window.BlobBuilder = function BlobBuilder() {
+      this.blob = new Object({ blobBuilder: true });
+    };
+
+    $httpBackend.expect('GET', audioRequestUrl).respond(201, 'Error');
+    expect(AssetsBackendApiService.isCached('myfile.mp3')).toBe(false);
+
+    AssetsBackendApiService.loadAudio('0', 'myfile.mp3').then(
+      successHandler, failHandler);
+    $httpBackend.flush();
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalledWith('myfile.mp3');
+    $httpBackend.verifyNoOutstandingExpectation();
+  });
+
+  it('should fetch file when Blob throws a TypeError and BlobBuilder' +
+    ' is correctly implemented', function() {
+    var successHandler = jasmine.createSpy('success');
+    var failHandler = jasmine.createSpy('fail');
+
+    spyOn(window, 'Blob').and.callFake(function() {
+      throw new TypeError();
+    });
+
+    window.BlobBuilder = function BlobBuilder() {
+      this.blob = new Object({ blobBuilder: true });
+      this.append = function append(data) {};
+      this.getBlob = function getBlob(contentType) {
+        return this.blob;
+      };
+    };
+
+    $httpBackend.expect('GET', audioRequestUrl).respond(201, 'Error');
+    expect(AssetsBackendApiService.isCached('myfile.mp3')).toBe(false);
+
+    AssetsBackendApiService.loadAudio('0', 'myfile.mp3').then(
+      successHandler, failHandler);
+    $httpBackend.flush();
+    expect(successHandler).toHaveBeenCalledWith(
+      audioFileObjectFactory.createNew(
+        'myfile.mp3',
+        { blobBuilder: true }
+      ));
+    expect(failHandler).not.toHaveBeenCalled();
+    $httpBackend.verifyNoOutstandingExpectation();
+  });
+
+  it('should successfully save an audio', function(done) {
+    var successMessage = 'Audio was successfully saved.';
+    // @ts-ignore
+    spyOn($, 'ajax').and.callFake(function() {
+      var d = $.Deferred();
+      d.resolve(successMessage);
+      return d.promise();
+    });
+
+    AssetsBackendApiService.saveAudio('0', 'a.mp3', new File([], 'a.mp3'))
+      .then(function(response) {
+        expect(response).toBe(successMessage);
+      }).then(done, done.fail);
+
+    // $q Promises need to be forcibly resolved through a JavaScript digest,
+    // which is what $apply helps kick-start.
+    $rootScope.$apply();
+  });
+
+  it('should not save an audio', function(done) {
```

Comment:
Can you please add the condition in which it should fail?

Reply:
Done.